### PR TITLE
Add prediction to api

### PR DESCRIPTION
**Pull request by DelaneySteve**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,6 @@
 requests >= 2.31.0
 fastapi==0.109.0
-uvicorn>=0.26.0, <1.0.0
\ No newline at end of file
+uvicorn>=0.26.0, <1.0.0
+pandas 
+scikit-learn
```
Comment (james-zafar) [inline requirements.txt]: Is it worth adding version constraints here?

File: src/api/operations/create_prediction.py [hunk 2]
```diff
@@ -31,20 +38,36 @@ def get_api_key(api_key_attempt: str = Security(api_key_header)) -> str:
 
 prediction_router = APIRouter(prefix='/air-quality')
 
-@prediction_router.post('', response_model=PostAirQualityResponse, status_code=201)
-async def predict_air_quality(airport: Airport, api_key: str = Security(get_api_key)) -> JSONResponse: # pylint: disable=unused-argument
 
+@prediction_router.on_event('startup')
```
Comment (james-zafar) [inline src/api/operations/create_prediction.py]: Did you decide against using lifespan events?

File: src/api/operations/create_prediction.py [hunk 3]
```diff
@@ -3,23 +3,30 @@
...
 
 # import API key from environment variables
-load_dotenv() # loads the variables from the .env file to the current session's environment
+load_dotenv()  # loads the variables from the .env file to the current session's environment
 API_KEY = str(os.getenv('API_KEY'))
 api_key_header = APIKeyHeader(name='X-API-key')
 
+MODEL_PATH = 'src/model/model_new.pickle'
+
+rf_model = Model()
```
Comment (james-zafar) [inline src/api/operations/create_prediction.py]: Why do we instantiate an empty model here?

File: src/api/resources/airport.py [hunk 4]
```diff
@@ -6,3 +6,12 @@ class Airport(BaseModel):
...
     total_arrivals: int
     total_departures: int
+
+    def to_json(self) -> dict[str, str | int | list[dict[str, int]]]:
+        return {
+            'altitude': self.altitude,
+            'runways': [runway.to_json() for runway in self.runways],
+            'total_arrivals': self.total_arrivals,
+            'total_departures': self.total_departures
+            }
```
Comment (james-zafar) [inline src/api/resources/airport.py]: Is this the same as the types specified in `types.Airport`?

File: src/api/operations/create_prediction.py (hunk 2 above)
Comment (DelaneySteve) [reply src/api/operations/create_prediction.py]: Yeah it wasn't working :')

File: README.md [hunk 5]
```diff
@@ -5,3 +5,8 @@ The AirPort Air Quality (APAQ) Predictor is an API tool that can be used to pred
 1. Create a virtual environment by executing `make` in the root.
 2. Activate the virtual environment by executing `. venv/Scripts/activate` (for Windows) or `. venv/bin/activate` (for Mac/Linux) also in the root directory.
 3. Run the program by executing `make run` in the root directory.
+
+## Train a Random Forest Model
+1. Execute: `python -m src.model.main --aug-airports-load-file DATA_FILE_PATH` where DATA_FILE_PATH is the local path to the training dataset
```
Comment (james-zafar) [inline README.md]: I don't really like the name `--aug-airports-load-file` - could we change this to something a bit more sensible?

File: src/api/operations/create_prediction.py [hunk 6]
```diff
@@ -2,24 +2,33 @@
...
 # import logger
 logger = setup_logger()
 
 # import API key from environment variables
-load_dotenv() # loads the variables from the .env file to the current session's environment
+load_dotenv()  # loads the variables from the .env file to the current session's environment
 API_KEY = str(os.getenv('API_KEY'))
 api_key_header = APIKeyHeader(name='X-API-key')
 
+MODEL_PATH = 'src/model/rf_model.pickle'
```
Comment (james-zafar) [inline src/api/operations/create_prediction.py]: Perhaps annotate this as `Final[str]` for clarity as it's a constent

File: src/api/operations/create_prediction.py [hunk 7]
```diff
@@ -2,24 +2,33 @@
...
+
+rf_model = None # type: Model # type: ignore
```
Comment (james-zafar) [inline src/api/operations/create_prediction.py]: You should only need one of `# type: Model ` and `# type: ignore`

File: src/model/main.py [hunk 8]
```diff
@@ -3,6 +3,8 @@
 
 from src.model.model import Model
 
+SAVE_MODEL_PATH = 'src/model/model_new.pickle'
```
Comment (james-zafar) [inline src/model/main.py]: I don't think this should be inside `src` as it's not part of the code - perhaps create a `./model` directory in the root instead

File: src/model/model.py [hunk 9]
```diff
@@ -11,18 +11,16 @@
 
 
 class Model:
-    def __init__(self) -> None:
-        self._model = None  # type: RandomForestRegressor
```
Comment (james-zafar) [inline src/model/model.py]: Why have you removed the init?

File: src/api/resources/airport.py [hunk 10]
```diff
@@ -8,3 +8,11 @@ class Airport(BaseModel):
     runways: list[Runway]
     total_arrivals: int
     total_departures: int
+
+    def to_json(self) -> dict[str, str | int | list[dict[str, int]]]:
```
Comment (james-zafar) [inline src/api/resources/airport.py]: Can you rename this to `json` to be consistent with the other methods in other types please?

File: src/api/resources/runway.py [hunk 11]
```diff
@@ -36,3 +36,9 @@ class RunwaySurface(Enum):
 class Runway(BaseModel):
     length_in_ft: int
     # surface: RunwaySurface # Surface parameter hidden until required by the model
+
+
+    def to_json(self) -> dict[str, int]:
```
Comment (james-zafar) [inline src/api/resources/runway.py]: Same here

File: src/model/model.py [hunk 12]
```diff
@@ -13,18 +13,21 @@
 
 
 class Model:
+    _model = None # type: RandomForestRegressor
```
Comment (james-zafar) [inline src/model/model.py]: Why do you want this to be an class variable?

File: src/model/model.py [hunk 13]
```diff
@@ -13,18 +13,21 @@
...
+
     def __init__(self) -> None:
-        self._model = None  # type: RandomForestRegressor
```
Comment (james-zafar) [inline src/model/model.py:17]: I think you should instantiate the model here?

File: src/model/model.py [hunk 14]
```diff
@@ -13,18 +13,21 @@
...
+        pass
 
-    def load_trained_model(self, filename: str) -> None:
+    @classmethod
+    def load_trained_model(cls, filename: str) -> None:
```
Comment (james-zafar) [inline src/model/model.py]: A `classmethod` should return an instance of `cls`

**Review by james-zafar — APPROVED** [review]
Looks good, thanks

**Pull request merged**